File: otherLangs/haskell/03.hs
import Control.Monad ()

main :: IO ()
main = do
  contents <- readFile "inputs/03.txt"
  let input = map (cycle . map (=='#')) $ lines contents
      slopes = [(1,1),(1,3),(1,5),(1,7),(2,1)]
  print $ solveSlope input (1,3)
  print $ product $ map (solveSlope input) slopes

solveSlope :: [[Bool]]  -> (Int,Int) -> Int
solveSlope ls (dy,dx) = length $ filter (\y -> ls !! y !! (y * dx `div` dy)) [dy, dy * 2 .. length ls - 1]

File: 2020/03.js
import { product, range, readFile } from '../tools.ts';

const input = readFile('./inputs/03.txt').split('\n').map(l => l.split('').map(c => c === '#'));

const solveSlope = ([xS, yS]) => 
  range(yS, input.length, yS)
    .filter(y => input[y][(y * xS / yS) % input[y].length])
    .length

console.log(solveSlope([3, 1]))

const slopes = [
  [1,1],
  [3,1],
  [5,1],
  [7,1],
  [1,2],
];

console.log(product(slopes.map(solveSlope)))

